lib/learn_test/git/wip/base.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'git'
+require 'logger'
+require 'tempfile'
+
+module LearnTest
+  module Git
+    module Wip
+      class Base < ::Git::Path
+        TEMPFILE = '.wip'
+        PREFIX = 'refs/wip/'
+
+        attr_reader :working_branch, :wip_branch
+
+        def initialize(base:, message:)
+          @base = base
+          @message = message
+          @success = nil
+
+          current_branch = @base.current_branch
+
+          raise NoCommitsError, 'master' if current_branch.nil? # TODO: Swap to `main`?
+
+          @tmp = Tempfile.new(TEMPFILE)
+          @working_branch = Branch.new(base: @base, name: current_branch)
+          @wip_branch = Branch.new(base: @base, name: "#{PREFIX}#{current_branch}")
+        end
+
+        def process!
+          if @wip_branch.last_revision
+            merge = @base.merge_base(@wip_branch.last_revision, @working_branch.last_revision)
+
+            @wip_branch.parent = if merge == @working_branch.last_revision
+                                   @wip_branch.last_revision
+                                 else
+                                   @working_branch.last_revision
+                                 end
+          else
+            @wip_branch.parent = @working_branch.last_revision
+          end
+
+          new_tree = build_new_tree(@wip_branch.parent)
+          @base.diff(new_tree, @wip_branch.parent)
+
+          # tree_diff = @base.diff(new_tree, @wip_branch.parent)
+          # raise NoChangesError, @wip_branch if tree_diff.count.zero?
```

Comment:
Does this not need to be here?

Reply:
Keeping it just until wider testing is complete.